projectq/setups/decompositions/phaseestimation_test.py

```diff
@@ -0,0 +1,231 @@
+#   Copyright 2017 ProjectQ-Framework (www.projectq.ch)
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"Tests for projectq.setups.decompositions.phaseestimation.py."
+
+import copy
+import cmath
+import numpy as np
+import pytest
+
+from projectq import MainEngine
+from projectq.backends import Simulator
+from projectq.cengines import (AutoReplacer, DecompositionRuleSet,
+                               DummyEngine, InstructionFilter, MainEngine)
+
+from projectq.ops import X, H, All, Measure, Tensor, Ph, StatePreparation
+
+from projectq.ops import (BasicGate)
+
+from projectq.ops import QPE
+from projectq.setups.decompositions import phaseestimation as pe
+from projectq.setups.decompositions import qft2crandhadamard as dqft
+import projectq.setups.decompositions.stateprep2cnot as stateprep2cnot
+import projectq.setups.decompositions.uniformlycontrolledr2cnot as ucr2cnot
+
+
+class PhaseX(BasicGate):
```

Comment:
These new gates should go into a separate PR.
This PR adds phase estimation only.

Reply:
I created these gates just for testing purposes, with no intention to be part of the framework. This is why I included them in the _test.py. Do you prefer I to eliminate them from the test.py (and eliminate the corresponding tests) or to maintain them in the test.py?

I'll wait for your response on this before recommit

Thanks again
Best

Reply:
Oh I see. Is there a reason why you'd like to test with these gates specifically?
I think it would be nice to use standard functionality in the tests so other users can use them to figure out how to use QPE in their own projects. Would you mind adapting the tests to use, e.g., `Rz` or `Rx`?
If the gates you have added are useful in other settings, we can also add them to `projectq.ops` in a later PR.
Thanks!

Reply:
The reason is to test nontrivial phases in case of PhaseX, and to test more than 1 quibit in system qubits in the case of PhaseXxX. I can easily change the PhaseX to use Ph, but in the case of PhaseXxX I'm not sure if I can easily use a standard resource. I don't think a gate like this can be useful for others outside of testing for PE.

I propose to change PhaseX to Ph and create a new PhxX for 2-qubit system-qubits just for testing purpouses without intention to add it to `projectq.ops`  later on, what do you think? 

Thank you

Reply:
Instead of testing a 2-qubit gate directly, you can also just test QPE by providing a function which does, e.g.,
```
CNOT | (q1, q2)
Rz(theta) | q2
CNOT | (q1, q2)
```
so you can remove both gates and use standard gates for all tests.

Thanks a lot!

Reply:
OK, great!

I'll change to use a 2-qubit  test using something like your proposal. Let me change the test and I'll recommit again (not before the weekend :-(  )

Thanks a lot!